File: src/main/java/ai/labs/eddi/configs/groups/model/GroupWorkspace.java
/*
 * Copyright EDDI contributors
 * SPDX-License-Identifier: Apache-2.0
 */
package ai.labs.eddi.configs.groups.model;

import java.time.Instant;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * A group's standing workspace (I13): the persistent half of a team. A group
 * <em>conversation</em> is an episode; the workspace is what survives between
 * episodes — the backlog, the cadences that pull work from it, and the team's
 * running metrics.
 * <p>
 * One document per group config id, in its own collection — deliberately NOT
 * embedded in the versioned {@link AgentGroupConfiguration}: the workspace
 * mutates on every cadence run and backlog write, and versioning it alongside
 * the config would turn each of those into a config revision.
 * <p>
 * <b>Concurrency.</b> Cadence fires are cluster-wide (any pod's poller may
 * claim one), so "is a cadence discussion already running?" is decided by a
 * conditional store write on {@link #runningDiscussionId} — see
 * {@code IGroupWorkspaceStore#claimRun} — never by in-JVM locks. The idle value
 * is the empty string, not {@code null}, because the conditional write compares
 * a stored field against a concrete value.
 * <p>
 * <b>Per-member stats are reliability RECORDING only</b> (research-adopted
 * substrate): nothing routes or weights on them in v1, and that is a deliberate
 * scope cut, not an oversight.
 *
 * @author ginccc
 */
public class GroupWorkspace {

    /**
     * Document shape version for THIS collection (independent of
     * {@code GroupConversation}'s). Version 1 is the first that ever existed.
     */
    public static final int CURRENT_SCHEMA_VERSION = 1;

    /**
     * Backlog ceiling. A backlog is a working set, not an archive — past this size,
     * adding tasks fails with an actionable error instead of growing unboundedly
     * (an LLM or a runaway integration can call add in a loop).
     */
    public static final int MAX_BACKLOG_SIZE = 200;

    /** Default number of backlog tasks one cadence run pulls. */
    public static final int DEFAULT_MAX_BACKLOG_TASKS_PER_RUN = 5;

    /** {@link #runningDiscussionId}'s idle value — see the class Javadoc. */
    public static final String NO_RUNNING_DISCUSSION = "";

    private String id;
    private int schemaVersion = CURRENT_SCHEMA_VERSION;
    private String groupId;
    /**
     * The team's persistent backlog. Reuses {@link SharedTaskList} whole —
     * statuses, dependency plumbing, caps — so a cadence run can hand pulled tasks
     * straight to the task-force machinery without a parallel task model.
     */
    private SharedTaskList backlog = new SharedTaskList();
    private WorkspaceMetrics metrics = new WorkspaceMetrics();
    private List<Cadence> cadences = new CopyOnWriteArrayList<>();
    /**
     * The cadence discussion currently in flight, or
     * {@link #NO_RUNNING_DISCUSSION}. Claimed by conditional write before a cadence
     * run starts; cleared by writeback once that discussion reaches a terminal
     * state.
     */
    private String runningDiscussionId = NO_RUNNING_DISCUSSION;
    /**
     * Backlog task ids the running discussion pulled — what writeback returns to
     * PENDING if the discussion fails, and matches outcomes against when it
     * completes.
     */
    private List<String> pulledTaskIds = new CopyOnWriteArrayList<>();
    private Instant created;
    private Instant lastModified;
    /**
     * Optimistic-concurrency stamp for read-modify-write surfaces (review finding:
     * two concurrent backlog adds could both pass the cap/duplicate checks against
     * their own snapshots and the later whole-document write dropped the earlier
     * task). Bumped by {@code IGroupWorkspaceStore.casRevision}; stored as a string
     * because the conditional-store primitive compares string field equality.
     * {@code null} on documents created before this field existed.
     */
    private String revision = "0";

    /**
     * One scheduled pull from the backlog.
     *
     * @param cadenceId
     *            stable id within this workspace
     * @param scheduleRef
     *            id of the {@code ScheduleConfiguration} driving this cadence —
     *            cadence execution rides {@code SchedulePollerService}'s
     *            cluster-aware claim/lease/retry machinery whole; the workspace
     *            never runs its own timer
     * @param inputTemplate
     *            optional Qute template rendered over the backlog summary to
     *            produce the discussion question; {@code null} uses a plain default
     * @param maxBacklogTasksPerRun
     *            how many executable tasks one fire pulls (default
     *            {@value #DEFAULT_MAX_BACKLOG_TASKS_PER_RUN})
     * @param maxCostPerRun
     *            dollar ceiling for one cadence discussion — dollar-primary per the
     *            Dream precedent; {@code null} falls back to the group's own
     *            {@code maxCostPerDiscussion}
     * @param createdBy
     *            principal that created the cadence — the identity its discussions
     *            run under (a cadence run must be attributable to whoever set it
     *            up, not to a synthetic scheduler user)
     */
    public record Cadence(String cadenceId, String scheduleRef, String inputTemplate, int maxBacklogTasksPerRun,
            Double maxCostPerRun, String createdBy) {

        public Cadence {
            if (maxBacklogTasksPerRun <= 0) {
                maxBacklogTasksPerRun = DEFAULT_MAX_BACKLOG_TASKS_PER_RUN;
            }
        }
    }

    /**
     * The team's running totals. A plain mutable POJO — writeback is the only
     * writer, and it runs under the workspace's single-writer claim.
     */
    public static class WorkspaceMetrics {

        private int discussions;
        private int tasksVerified;
        private double totalCost;
        private Instant lastRunAt;
        private Map<String, MemberStats> perMemberStats = new ConcurrentHashMap<>();

        public int getDiscussions() {
            return discussions;
        }

        public void setDiscussions(int discussions) {
            this.discussions = discussions;
        }

        public int getTasksVerified() {
            return tasksVerified;
        }

        public void setTasksVerified(int tasksVerified) {
            this.tasksVerified = tasksVerified;
        }

        public double getTotalCost() {
            return totalCost;
        }

        public void setTotalCost(double totalCost) {
            this.totalCost = totalCost;
        }

        public Instant getLastRunAt() {
            return lastRunAt;
        }

        public void setLastRunAt(Instant lastRunAt) {
            this.lastRunAt = lastRunAt;
        }

        public Map<String, MemberStats> getPerMemberStats() {
            return perMemberStats;
        }

        public void setPerMemberStats(Map<String, MemberStats> perMemberStats) {
            this.perMemberStats = perMemberStats != null
                    ? new ConcurrentHashMap<>(perMemberStats)
                    : new ConcurrentHashMap<>();
        }
    }

    /**
     * Reliability recording for one member — see the class Javadoc's scope note.
     */
    public static class MemberStats {

        private int tasksVerified;
        private int tasksFailed;

        public int getTasksVerified() {
            return tasksVerified;
        }

        public void setTasksVerified(int tasksVerified) {
            this.tasksVerified = tasksVerified;
        }

        public int getTasksFailed() {
            return tasksFailed;
        }

        public void setTasksFailed(int tasksFailed) {
            this.tasksFailed = tasksFailed;
        }
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public int getSchemaVersion() {
        return schemaVersion;
    }

    public void setSchemaVersion(int schemaVersion) {
        this.schemaVersion = schemaVersion;
    }

    public String getGroupId() {
        return groupId;
    }

    public void setGroupId(String groupId) {
        this.groupId = groupId;
    }

    public SharedTaskList getBacklog() {
        return backlog;
    }

    public void setBacklog(SharedTaskList backlog) {
        this.backlog = backlog != null ? backlog : new SharedTaskList();
    }

    public WorkspaceMetrics getMetrics() {
        return metrics;
    }

    public void setMetrics(WorkspaceMetrics metrics) {
        this.metrics = metrics != null ? metrics : new WorkspaceMetrics();
    }

    /**
     * Read-only view — mutation goes through {@link #addCadence} and
     * {@link #removeCadence}, so the schedule references cannot be edited behind
     * the workspace's back.
     */
    public List<Cadence> getCadences() {
        return Collections.unmodifiableList(cadences);
    }

    public void addCadence(Cadence cadence) {
        if (cadence != null) {
            cadences.add(cadence);
        }
    }

    /** @return {@code true} if a cadence with that id existed and was removed */
    public boolean removeCadence(String cadenceId) {
        return cadences.removeIf(c -> c.cadenceId().equals(cadenceId));
    }

    public void setCadences(List<Cadence> cadences) {
        this.cadences = cadences != null ? new CopyOnWriteArrayList<>(cadences) : new CopyOnWriteArrayList<>();
    }

    public String getRunningDiscussionId() {
        return runningDiscussionId;
    }

    public void setRunningDiscussionId(String runningDiscussionId) {
        this.runningDiscussionId = runningDiscussionId != null ? runningDiscussionId : NO_RUNNING_DISCUSSION;
    }

    public List<String> getPulledTaskIds() {
        return pulledTaskIds;
    }

    public void setPulledTaskIds(List<String> pulledTaskIds) {
        this.pulledTaskIds = pulledTaskIds != null ? new CopyOnWriteArrayList<>(pulledTaskIds) : new CopyOnWriteArrayList<>();
    }

    public Instant getCreated() {
        return created;
    }

    public void setCreated(Instant created) {
        this.created = created;
    }

    public Instant getLastModified() {
        return lastModified;
    }

    public void setLastModified(Instant lastModified) {
        this.lastModified = lastModified;
    }

    public String getRevision() {
        return revision;
    }

    public void setRevision(String revision) {
        this.revision = revision;
    }
}

File: ui/manager/src/lib/api/group-workspace.ts
import { api } from "../api-client";
import type { SharedTaskList, TaskItem } from "./groups";

// ─────────────────────────────────────────────────────────────────
// I13 — standing teams: a persistent workspace (backlog + metrics +
// cadences) bolted onto a group, run unattended by a scheduler.
// ─────────────────────────────────────────────────────────────────

/** Backend `GroupWorkspace.MAX_BACKLOG_SIZE`. */
export const WORKSPACE_MAX_BACKLOG_SIZE = 200;
/** Backend `GroupWorkspace.DEFAULT_MAX_BACKLOG_TASKS_PER_RUN`, used when a cadence's own value is non-positive. */
export const WORKSPACE_DEFAULT_MAX_BACKLOG_TASKS_PER_RUN = 5;
/** Backend `RestGroupWorkspace.MAX_CADENCES_PER_WORKSPACE`. */
export const WORKSPACE_MAX_CADENCES = 20;
/** Backend `RestGroupWorkspace.MAX_INPUT_TEMPLATE_LENGTH`. */
export const WORKSPACE_MAX_INPUT_TEMPLATE_LENGTH = 4000;
/** Backend `SharedTaskList.MAX_AGENT_TASK_SUBJECT_LENGTH` — same cap the addBacklogTask endpoint enforces. */
export const WORKSPACE_MAX_TASK_SUBJECT_LENGTH = 200;
/** Backend `SharedTaskList.MAX_AGENT_TASK_DESCRIPTION_LENGTH`. */
export const WORKSPACE_MAX_TASK_DESCRIPTION_LENGTH = 4000;

/** Sentinel `runningDiscussionId` value meaning "no cadence run is currently in flight" — backend `GroupWorkspace.NO_RUNNING_DISCUSSION`. */
export const NO_RUNNING_DISCUSSION = "";

/**
 * A recurring, scheduled discussion that pulls from the team's own backlog
 * (I13). Its fire-time state (next/last fire, enabled) lives on a PAIRED
 * `ScheduleConfiguration` row in the SAME generic schedule store used by Dream
 * consolidation and plain cron triggers — discriminated by
 * `metadata.teamCadenceType === "team_cadence"` — never a bespoke schedule type.
 * That row is read-only observability for this UI; every mutation must go
 * through `addWorkspaceCadence`/`deleteWorkspaceCadence` below, never the
 * generic `/schedulestore/schedules` endpoints, or `GroupWorkspace.cadences` and
 * the schedule store fall out of sync.
 */
export interface Cadence {
  cadenceId: string;
  /** Id of the paired row in the generic `/schedulestore/schedules` collection. */
  scheduleRef: string;
  /** Qute template rendered over the pulled backlog tasks at fire time. `null` falls back to a plain task listing. */
  inputTemplate: string | null;
  /** Top-N backlog tasks (by priority, descending) pulled per fire. Non-positive falls back to the default (5). */
  maxBacklogTasksPerRun: number;
  /** Dollar ceiling for one fire's discussion. `null` defers entirely to the group's own `protocol.maxCostPerDiscussion` — when both are set, the tighter of the two wins. */
  maxCostPerRun: number | null;
  /** The principal every fire of this cadence runs the discussion as — fixed at cadence-creation time, not the identity of whoever is currently viewing the workspace. */
  createdBy: string;
}

/** Per-member reliability counters (I13) — RECORDING only; nothing in v1 routes, weights, or filters assignment on these numbers. */
export interface MemberStats {
  tasksVerified: number;
  tasksFailed: number;
}

/** Aggregate stats for a standing team's workspace (I13), updated after every cadence fire settles (success or failure). */
export interface WorkspaceMetrics {
  discussions: number;
  tasksVerified: number;
  totalCost: number;
  lastRunAt: string | null;
  perMemberStats: Record<string, MemberStats>;
}

/**
 * A group's standing-team workspace (I13): a persistent backlog + metrics +
 * cadence list, auto-created empty on first `GET` if the group has none yet.
 * `backlog` here is the FULL `SharedTaskList` shape (`{tasks, awardedBids}`) —
 * contrast with {@link getGroupWorkspaceBacklog}, which returns a bare array.
 */
export interface GroupWorkspace {
  id: string;
  schemaVersion: number;
  groupId: string;
  backlog: SharedTaskList;
  metrics: WorkspaceMetrics;
  cadences: Cadence[];
  /** Id of the discussion a cadence fire currently has claimed, or `""` (see {@link NO_RUNNING_DISCUSSION}) when idle. At most one cadence run in flight per workspace, even with multiple cadences. */
  runningDiscussionId: string;
  /** Backlog task ids the in-flight claim pulled — cleared when the run settles. */
  pulledTaskIds: string[];
  created: string;
  lastModified: string;
  revision: string;
}

/** Body of `addWorkspaceBacklogTask`. */
export interface AddBacklogTaskRequest {
  subject: string;
  description?: string;
  /** Higher runs earlier when a cadence pulls — NOT the "0 = highest" convention `TaskItem`'s general Javadoc describes elsewhere. */
  priority?: number;
}

/** Body of `addWorkspaceCadence`. */
export interface AddCadenceRequest {
  cronExpression: string;
  /** IANA zone id. Blank/absent defaults to `"UTC"`. */
  timeZone?: string;
  inputTemplate?: string;
  maxBacklogTasksPerRun?: number;
  maxCostPerRun?: number;
}

/**
 * Read a group's standing-team workspace, auto-creating an empty one if none
 * exists yet. Read-repairs first: if a previously claimed cadence discussion has
 * since finished (COMPLETED/FAILED/CANCELLED), its outcome is written back to
 * the backlog before the response is built — so this call can itself mutate
 * task statuses and metrics as a side effect of reading them.
 * GET /groupstore/groups/{groupId}/workspace
 */
export function getGroupWorkspace(groupId: string): Promise<GroupWorkspace> {
  return api.get<GroupWorkspace>(
    `/groupstore/groups/${encodeURIComponent(groupId)}/workspace`,
  );
}

/**
 * Read just the backlog task array (same read-repair as {@link getGroupWorkspace}).
 * NOTE the shape difference: this returns a bare `TaskItem[]`, not the
 * `{tasks, awardedBids}` envelope `workspace.backlog` carries.
 * GET /groupstore/groups/{groupId}/workspace/backlog
 */
export function getGroupWorkspaceBacklog(groupId: string): Promise<TaskItem[]> {
  return api.get<TaskItem[]>(
    `/groupstore/groups/${encodeURIComponent(groupId)}/workspace/backlog`,
  );
}

/**
 * File a new backlog task. Subject must be non-blank, unique (case-insensitive)
 * within the backlog, and the backlog must be under its 200-task cap.
 * POST /groupstore/groups/{groupId}/workspace/backlog
 * Failures: 400 (blank/oversized subject or description), 409 (backlog full,
 * duplicate subject, or lost a concurrent-write retry — the thrown ApiError's
 * `message` names which).
 *
 * NOTE: there is no companion delete/update endpoint for a single backlog task
 * (confirmed absent on both REST and MCP) — the only way a task leaves PENDING
 * today is via a cadence run pulling and completing it.
 */
export function addWorkspaceBacklogTask(
  groupId: string,
  request: AddBacklogTaskRequest,
): Promise<TaskItem> {
  return api.post<TaskItem>(
    `/groupstore/groups/${encodeURIComponent(groupId)}/workspace/backlog`,
    request,
  );
}

/**
 * Add a recurring cadence. Creates the paired `ScheduleConfiguration` row first;
 * if appending the `Cadence` to the workspace then fails, the just-created
 * schedule is deleted as compensation (server-side — nothing for a caller to
 * clean up either way).
 * POST /groupstore/groups/{groupId}/workspace/cadences
 * Failures: 400 (blank/unparseable cron, oversized inputTemplate), 409 (already
 * at the 20-cadence cap).
 */
export function addWorkspaceCadence(
  groupId: string,
  request: AddCadenceRequest,
): Promise<Cadence> {
  return api.post<Cadence>(
    `/groupstore/groups/${encodeURIComponent(groupId)}/workspace/cadences`,
    request,
  );
}

/**
 * Delete a cadence — removes its paired schedule row first (best-effort; a
 * schedule-delete failure does not block the cadence removal), then the
 * `Cadence` entry itself.
 * DELETE /groupstore/groups/{groupId}/workspace/cadences/{cadenceId}
 * 204 on success. Failures: 404 (no workspace, or no such cadence).
 */
export function deleteWorkspaceCadence(
  groupId: string,
  cadenceId: string,
): Promise<void> {
  return api.delete(
    `/groupstore/groups/${encodeURIComponent(groupId)}/workspace/cadences/${encodeURIComponent(cadenceId)}`,
  );
}
